[1단계 - Virtual DOM 만들기] 체프(심문성) 미션 제출합니다.

Puterism opened the pull request:
하루 안녕하세요! 체프입니다
여차저차 필수 요구사항을 충족할 수 있었습니다
JSX Parser까지 하면 좋았을텐데 아쉽네요 ㅠㅠ
조금이라도 거슬리는 부분이 있으시다면 가감없는 피드백 부탁드립니다

![vdom-mission (1)](https://user-images.githubusercontent.com/2542730/137751222-69e786eb-2a1d-4ebf-aeb1-aada7fed3732.gif)

## 구현한 VDOM의 특징과 한계점

- 초기 렌더 시에만 `render` 함수를 사용하고, 업데이트할 때는 `update` 함수를 사용합니다.
- 이벤트는 `on`으로 시작하는 속성 이름이면서, `value`가 function일 경우에만 `addEventListener`가 실행됩니다.
- 요소의 이름, 속성의 key, value의 변화, child의 변화를 감지하고 재렌더링합니다.
  - 이 때, 초기 렌더 시에 이벤트 바인딩한 `on`으로 시작하는 속성의 변화는 감지하지 못합니다.
- 상위 노드에서부터 하위 노드까지 Real DOM의 노드와 업데이트할 Virtual DOM의 노드를 비교하면서 변화를 감지합니다.
  - 만약, 상위 노드에서 변화가 감지되었다면, 해당 노드와 자식 노드들까지 전부 업데이트합니다.
  - 변화가 감지되지 않았다면, 렌더하지 않습니다.
- element나 attribute의 **개수 변화**는 올바르게 감지하지 못합니다.
  - 따라서, 초기 렌더 시의 동일한 개수의 element와 attribute로 update 해주어야 합니다.
- 변경된 부분을 감지할 때마다 렌더링이 동작하기 때문에, 변화를 한번에 모아서 렌더링하는 로직은 구현하지 못했습니다.
- 상태 관리를 별도로 구현하지 않았습니다. 버튼을 클릭하면 새로운 VDOM을 생성해서 update해주는 방식으로 만들었습니다. 

### 필수 요구사항

- [x] VirtualDOM 생성 팩토리 함수 구현 (aka createElement)
- [x] 렌더러 구현
- [x] 직접 구현한 렌더러와 VirtualDOM으로 Counter App 렌더링
- [x] DOM Node 비교
- [x] 업데이트 DOM Node
- [x] 렌더링한 Counter App 동작시키기

### 심화 요구사항

- [ ] 문자열 기반의 JSX Parser 구현
- [ ] AST 처리 기반의 JSX Parser 구현
- [ ] 동일한 Element인 경우 Key 속성을 만들어 순서를 지정하도록 구현
- [ ] Router 구현 후 Query Parameter를 사용해 초기 값 정의

Commit:
feat: add requirements

Commit:
feat: add requirements

Puterism committed:
feat: `createElement` 팩토리 함수로 생성한 Virtual Node로 Real DOM에 렌더하는 기능 구현

Puterism committed:
feat: VDOM을 업데이트하여 새롭게 다시 렌더하는 기능 구현

- 카운터 기능 동작

Puterism committed:
feat: 기존 VNode와 새로 업데이트하려는 VNode 객체가 동일한 경우, DOM 업데이트를 하지 않는 기능 구현

Puterism committed:
feat: Virtual DOM Node와 Real DOM Node를 비교하여, 바뀐 텍스트 노드만 렌더되도록 구현

- 텍스트 노드만 변경되고, 다른 부분이 변경된 경우는 아직 다시 렌더되도록 처리하지 않음

- 요소 이름이나 속성이 변경되었을 경우 `console.log` 메시지를 출력하도록 구현

- 모든 VNode의 children이 배열이 되도록 구현

Puterism committed:
chore: 주석 수정 및 DOM이 업데이트되었을 때 `console.log`가 실행되도록 수정

Puterism committed:
refactor: 사용하지 않는 key 속성  추가 로직 삭제

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+  eventList = [...eventList, { node, eventType, callback }];
+}
+
+function createRealDOMNode(vNode) {
```
Virtual의 약자로서 uppercase로 작성한 V가, `vNode`로 써있어서 쵸-큼 어색해 보였어연,,,
그치만 `VNode`는 클래스, `vNode`는 그걸로 찍어낸 인스턴스겠지라는 예상은 충분히 할 수 있어서 역시 괜찮다고 생각합니다.

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+function createRealDOMNode(vNode) {
+  // VNode가 아닌 값이 들어왔다면, 텍스트 노드로 생성한다.
```
친절한 주석. 편-안 ☕️

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+        const eventType = key.slice(2);
+
+        bindEvent(node, eventType, value);
+      }
```
이런식으로 리스너를 달아줄 수 있군요. 좋은 방법이네요. 👍👍👍

365kim replied:
스쳐지나가도 괜찮은 소소한 아이디어☺️
체프의 코드를 보며 그냥 생각난 아이디어인데 지원하는 모든 on-* 를 저장해두고(정말? 🙄), 유효한 핸들러 이름인지 한번 검증해주는 방법도 있을 거 같아요.

Puterism replied:
https://developer.mozilla.org/ko/docs/Web/API/GlobalEventHandlers
여기에서 다 끌어오면 되니까 가장 좋은 방법이라고 생각해요!

365kim replied:
오우 긁-긁

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+  if (attributes) {
+    // 모든 속성을 순회한다.
+    Object.entries(attributes).forEach(([key, value]) => {
+      const isBindEvent = key.slice(0, 2) === 'on';
```
스쳐지나가도 괜찮은 소소한 제안☺️
`isBindEvent` 도 좋지만 `isEventHandler`는 어떨까용?

Puterism replied:
더 적절한 네이밍이라고 생각해요! 제안 감사드립니다

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+      // value가 문자열이면, setAttribute한다.
+      else if (typeof value === 'string') {
+        node.setAttribute(key, value);
+      }
```
혹시 value의 type이 `boolean` 이거나 `number`인 경우에는 어떻게 되는걸까요?

`<input type="checkbox" checked>` 같은 경우에 value가 `undefined`일 것 같은데
cheffDOM에서는 요런 상황은 없는지 궁금합니다~~~

Puterism replied:
그 부분은 염두에 두지 못했네요... 말씀해주신대로 undefined가 될 것 같아요
짚어주셔서 감사드립니다

Puterism replied:
우선 무조건적으로 value를 받게 처리해주는 게 좋다고 판단했습니다.
createElement할 때, 속성도 객체 구조로 입력해주어야 해서, `checked`의 경우 `{ checked: true }` 의 형태로 입력받게끔 처리하기로 했습니다. 저 조건문에서 boolean도 받을 수 있도록 처리해보겠습니다. 피드백 감사합니다!

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+
+  targetNode.parentNode.replaceChild(fragment, targetNode);
+  console.log('DOM 업데이트됨');
+}
```
지극히 소소한 제안 ☺️
```suggestion
function updateRealDOMNode(vNode, targetNode) {
  const fragment = document.createDocumentFragment();
  const node = createRealDOMNode(vNode);
  
  fragment.appendChild(node);
  targetNode.parentNode.replaceChild(fragment, targetNode);
  console.log('DOM 업데이트됨');
}
```

Puterism replied:
ㅋㅋㅋㅋㅋㅋ 저도 제안해주신 게 더 좋다고 생각해요!

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+  const node = createRealDOMNode(vNode);
+  fragment.appendChild(node);
+
+  targetNode.parentNode.replaceChild(fragment, targetNode);
```
`replacedNode = parentNode.replaceChild(newChild, oldChild);`
저는 innerHTML을 빈문자열로 넣어주면서 비워줬는데,,, 훨씬 우아한 방식이네요! 배워갑니다 👏👏👏👏

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+// 루트 노드부터 시작해 상위 노드를 순서대로 쭉 비교한다.
+// 상위 노드가 변경되었으면, 하위 노드들도 변경된 상태로 간주한다.
+function diff(node, vNode) {
+  // node가 텍스트 노드일 때, vNode가 VNode가 아닌 다른 타입이라면
```
스쳐지나가도 괜찮은 소소한 제안☺️
다른 주석들은 읽음으로써 이해에 더 도움이 되었는데 이 문장 66라인의 코드와 완죠니 동일한 내용이어서 생략되어도 될 것 같아요.

Puterism replied:
똑같은 내용의 주석을 작성해뒀었군요 ㅠㅠ
피드백 감사합니다!

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+// 상위 노드가 변경되었으면, 하위 노드들도 변경된 상태로 간주한다.
+function diff(node, vNode) {
+  // node가 텍스트 노드일 때, vNode가 VNode가 아닌 다른 타입이라면
+  if (node instanceof Text && !(vNode instanceof VNode)) {
```
`instanceof Text`로 textNode인지 판단할 수 있군요?! 체프 덕분에 새로 배워가는 게 많네요 😄

Puterism replied:
instanceof로 하나하나 입력해보면서 찾아봤습니다... ㅋㅋㅋㅋㅋㅋ

365kim replied:
`노력형 천-재`

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+    }
+  }
+
+  // TODO: node.childNodes와 vNode.children의 길이가 다를 때, 참조 에러 막기
+  for (let i = 0; i < node.childNodes.length; i++) {
+    const currentNode = node.childNodes[i];
+    const currentVNode = vNode.children[i];
+
+    if (
+      currentNode.nodeName.toLowerCase() !== currentVNode.nodeName.toLowerCase()
+    ) {
+      console.log(
+        '노드 이름 다름',
+        currentNode.nodeName,
+        currentVNode.nodeName
+      );
+
+      updateRealDOMNode(currentVNode, currentNode);
+      continue;
+    }
```
스쳐지나가도 괜찮은 소소한 제안☺️
요번 미션에서 TODO를 해결하신다면
아래와 같이 더 긴 length 기준으로 순회하면서 if 조건문을 넣어주는 방식은 어떨까 싶습니당 :)

```js
const iterationLength = Math.max(node.childNodes.length, vNode.children.length);

for (let i = 0; i < iterationLength; i++) {
    if (i >= vNode.children.length) {
        // 리얼돔에서 사라져야할 노드 처리 로직 추가
        continue;
    } else if (i >= node.childNodes) {
      // 반대로 리얼돔에는 없지만 새로 붙여줘야할 노드들이 있다면 처리
        continue;
    }
    const currentNode = node.childNodes[i];
    const currentVNode = vNode.children[i];
   ...
   }

```

Puterism replied:
제안 감사드립니다!
위 코드를 활용해서 비교하는 과정에서 요소와 속성이 추가/삭제되는 부분까지는 구현했는데
이벤트를 관리하는 부분이 쉽지 않네요 😭
시간을 좀 더 들여야할 것 같아서 이 부분은 다음 미션에서 이어 나가보겠습니다 ㅠㅠ

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+    throw new Error(ERROR_MESSAGE.RENDER.EMPTY_ROOT_ELEMENT);
+  }
+
+  currentRootElement = rootElement;
```
`currentRootElement`와 `rootElement`는 역할이 어떻게 다른가요?

Puterism replied:
`rootElement`는 `render` 메소드에서 받는 인자로, 초기 렌더 시에 렌더될 요소를 받습니다.
이 요소를 저장해두었다가, DOM이 `update`되는 상황에서 업데이트될 VDOM과 비교하기 위해서 사용하는 것이 `currentRootElement`입니다.

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+  currentRootElement = rootElement;
+  currentRootVNode = vNode;
+
+  const fragment = document.createDocumentFragment();
```
혹시 `document.createDocumentFragment()`로 fragment를 생성하는 것과 `new DocumentFragment`로 생성하는 것의 차이점을 아시나용 👉👈 (몰라서 슬쩍 물어보는 질문이랍니당)

Puterism replied:
저도 잘 모르겠어서 검색해보고 왔습니다

https://developer.mozilla.org/ko/docs/Web/API/DocumentFragment#%EC%82%AC%EC%9A%A9%EB%B2%95

> 빈 DocumentFragment는 document.createDocumentFragment() 메서드나 생성자를 이용하여 만들 수 있습니다.

위와 같이 설명되어 있는 것을 보니, 두 방법의 차이점은 없는 것 같아요
저는 `createDocumentFragment`와 같이 별도로 제공되는 메소드가 있으면, 이를 사용하는 편이예요.

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+      const currentNodeChild = currentNode.childNodes[j];
+      const currentVNodeChild = currentVNode.children[j];
+
+      diff(currentNodeChild, currentVNodeChild);
```
재귀의 달인 👍

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+function update(vNode) {
+  if (typeof vNode !== 'object') {
```
`!(vNode instanceof VNode)`로 검증해주시는 경우와 `typeof vNode !== 'object'`로 검증해주시는 경우가 있는 것 같은데, 두 경우에 어떤 의도하신 차이점이 있을까용?

Puterism replied:
앗, 이건 제 실수입니다. `VNode`를 나중에 추가하는 과정에서 수정했어야 하는 코드입니다 ㅠㅠ
짚어주셔서 감사합니다!

365kim commented on constants.js:
```diff
@@ -0,0 +1,13 @@
+    INVALID_PARAMETER_TYPE_IS_EQUAL:
+      'isEqual 메소드의 인자는 VNode 객체여야 합니다.',
+  },
+};
```
구현만으로 벅찬데 에러처리까지... 👏👏👏
에러처리를 '기본'이라고 여기시는 것 같아 넘 멋지고 저도 본받아야겠다는 생각을 했습니당 👍

365kim commented on constants.js:
```diff
@@ -0,0 +1,13 @@
+  RENDER: {
+    EMPTY_ROOT_ELEMENT:
+      'rootElement가 비어있습니다. 렌더할 HTML 요소를 parameter에 할당해주세요.',
+    ONLY_CALL_ONCE: 'render는 한 번만 호츨할 수 있습니다.',
```
- 오타방지 위원회에서 나왔습니다.
```suggestion
    ONLY_CALL_ONCE: 'render는 한 번만 호출할 수 있습니다.',
```

Puterism replied:
캄사합니다

365kim commented on cheffeDOM.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+function render(vElement, rootElement) {
+  const vNode = vElement();
```
`vElement()`는 이름만 봐서는 호출될 거라고(함수일 거라고) 예상을 못했는데, App을 받고 있던 것이군요...!
인자 넣어줄 때(render 호출할 때) 아예 `App()`로 넣어주는 건 어떨까요..?

Puterism replied:
하루 의견대로 `App()`으로 넣어주면 `render` 메소드에서 함수일 것이라고 예상하기 힘든 문제를 해소할 수 있겠네요
피드백 감사드립니다!

365kim commented on index.js:
```diff
@@ -0,0 +1,135 @@
+let count = 0;
+
+const App = () => {
+  const onDecrease = () => {
+    count -= 1;
+
+    CheffeDOM.update(
+      Cheffe.createElement('div', { class: 'container' }, [
+        Cheffe.createElement('span', { class: 'count' }, [count]),
+        Cheffe.createElement('div', { class: 'btn-group' }, [
+          Cheffe.createElement(
+            'button',
+            {
+              onclick: onDecrease,
+            },
+            [Cheffe.createElement('strong', null, ['-'])]
+          ),
+          Cheffe.createElement(
+            'button',
+            {
+              onclick: onReset,
+            },
+            [Cheffe.createElement('strong', null, ['RESET'])]
+          ),
+          Cheffe.createElement(
+            'button',
+            {
+              onclick: onIncrease,
+            },
+            [Cheffe.createElement('strong', null, ['+'])]
+          ),
+        ]),
+      ])
+    );
+  };
```
`onDecrease`, `onIncrease`, `onReset`에서 호출하는 CheffeDOM.update의 틀린 그림 찾기를 시도해보았지만 실패했어요..! 다른 `on-*`가 추가되면 또 똑같이 써주어야하니까, 이 부분만 분리해서 코드 중복을 줄이면 어떨까 싶습니당 :)

Puterism replied:
상태가 변경되었을 때 `createElement`를 이용하여 새로운 VDOM을 다시 생성하고 다시 업데이트한다는 컨셉으로 코드를 구성했습니다. 상태 관리를 아직 구현하지 않아서, 저 생성코드를 별도로 분리하기에는 애매한 것 같습니다 ㅠㅠ

365kim replied:
앗, 10행부터 37행까지의 분리를 말씀드렸던건데, 저 부분은 분리해서 호출해도 지금이랑 똑같이 `createElement`를 호출해서 다시 생성하고 업데이트 되는게 맞지 않나용,,,? 그치만 2단계 진행하시면서 생각해두신게 있는 것 같군용,,,! 🙆🏻‍♀️🙆🏻‍♀️🙆🏻‍♀️

365kim reviewed (changes requested):
체프! ☕️

어려운 미션인데 멋지게 구현해주셨네용! 버추얼돔 단계별로 많이 고민하신게 느껴져서 코드를 읽는 것만으로도 많이 배울 수 있었어요. 👍 에러처리까지 해주신 것도 넘 멋지달까,,, (휴먼유조체)

PR 바디에 구현방향을 상세하게 적어주시고, 코드 레벨에도 주석을 친절하게 작성해주셔서 코드를 빠르게 파악할 수 있었어요. 저를 배려해주시는게 느껴져서 감사했어요 💙

> 상위 노드에서부터 하위 노드까지 Real DOM의 노드와 업데이트할 Virtual DOM의 노드를 비교하면서 변화를 감지합니다. 만약, 상위 노드에서 변화가 감지되었다면, 해당 노드와 자식 노드들까지 전부 업데이트합니다. 변화가 감지되지 않았다면, 렌더하지 않습니다.

V.DOM prev vs V.DOM next의 비교가 아니라, V.DOM vs R.DOM의 비교를 수행해주셨는데 Angular의 Incremental DOM 방식과 유사한 것 같아 관련 아티클 첨부해봅니당 :)

[Incremental DOM과 Virtual DOM 비교](https://ui.toast.com/weekly-pick/ko_20210819)

자세한 리뷰는 코드라인 별로 남겼어요.
시간 괜찮으실 때 확인 부탁드릴게용 :)

bigsaigon333 commented:
> [Incremental DOM과 Virtual DOM 비교](https://ui.toast.com/weekly-pick/ko_20210819)

좋은 아티클 공유 감사합니다 :+1:

Puterism committed:
chore: 메시지 오타 수정

Co-authored-by: 365kim <60066472+365kim@users.noreply.github.com>

Puterism committed:
refactor: 코드 리뷰 반영

- `isBindEvent` -> `isEventHandler`로 네이밍 변경

- VNode를 판별하는 조건문을 더 명확하게 수정

- 불필요한 주석 삭제

Puterism committed:
refactor: boolean, number으로 요소 속성을 받을 수 있도록 수정

Puterism commented:
하루 리뷰 감사합니다!
코드 리뷰 반영하고 다시 리뷰 요청 보내드립니다!

업데이트할 VDOM과 Real DOM과 비교하는 것이 더 확실하겠다는 생각이었는데
이는 말씀해주신 대로 Incremental DOM에 더 가깝네요 🥲 글 공유 감사드립니다!

Puterism committed:
refactor: `render` 메소드가 `vElement`의 반환값을 받도록 수정

365kim reviewed (approved):
체프,
피드백 반영해주셔서 감사해요~! 2단계 때도 좋은 코드 기대할게요 :) 👍👍👍
